Stratis.Bitcoin.Features.Miner/PosMinting.cs

```diff
@@ -626,97 +629,72 @@ private bool SignBlock(List<StakeTx> stakeTxes, Block block, ChainedBlock pindex
+                        continue;
                     }
 
                     // Do not add additional significant input.
-                    if (cointrx.TxOut.Value >= GetStakeCombineThreshold())
+                    if (stakeTx.TxOut.Value >= GetStakeCombineThreshold())
                     {
-                        this.logger.LogTrace("UTXO '{0}/{1}' rejected because its value is too big.", cointrx.OutPoint.Hash, cointrx.OutPoint.N);
+                        this.logger.LogTrace("UTXO '{0}/{1}' rejected because its value is too big.", stakeTx.OutPoint.Hash, stakeTx.OutPoint.N);
                         continue;
                     }
 
                     // Do not add input that is still too young.
-                    if (BlockValidator.IsProtocolV3((int)txNew.Time))
+                    // V3 case is properly handled by selection function.
+                    if (!BlockValidator.IsProtocolV3((int)coinstakeTx.Time))
                     {
-                        // Properly handled by selection function.
-                    }
-                    else
-                    {
-                        if (nTimeWeight < BlockValidator.StakeMinAge)
+                        if (timeWeight < BlockValidator.StakeMinAge)
                             continue;
                     }
 
-                    txNew.Inputs.Add(new TxIn(new OutPoint(cointrx.UtxoSet.TransactionId, cointrx.OutputIndex)));
+                    coinstakeTx.Inputs.Add(new TxIn(new OutPoint(stakeTx.UtxoSet.TransactionId, stakeTx.OutputIndex)));
 
-                    nCredit += cointrx.TxOut.Value;
-                    vwtxPrev.Add(coin);
+                    coinstakeInputsValue += stakeTx.TxOut.Value;
+                    coinstakeInputs.Add(stakeTx);
 
                     this.logger.LogTrace("UTXO '{0}/{1}' joined to coinstake transaction.");
+
+                    // Stop adding more inputs if already too many inputs.
+                    if (coinstakeTx.Inputs.Count >= 100)
+                    {
+                        this.logger.LogTrace("Number of coinstake inputs reached the limit of 100.");
+                        break;
+                    }
                 }
             }
 
             // Calculate coin age reward.
-            // TODO: This is not used  anywhere, delete it?
-            ulong nCoinAge;
-            if (!this.posConsensusValidator.StakeValidator.GetCoinAge(this.chain, this.coinView, txNew, pindexPrev, out nCoinAge))
-                return false; //error("CreateCoinStake : failed to calculate coin age");
+            if (!this.posConsensusValidator.StakeValidator.GetCoinAge(this.chain, this.coinView, coinstakeTx, chainTip, out ulong coinAge))
```

Comment:
FYI for your knowledge coin age is not use in this version of POS anymore, so I am not sure if this condition is required (I think its not checked in consensus)

For reference:
https://blackcoin.co/blackcoin-pos-protocol-v2-whitepaper.pdf

Reply:
Yes I agree with you but did not actually put any serious time into exploring if i can completely remove that. I'll create issue